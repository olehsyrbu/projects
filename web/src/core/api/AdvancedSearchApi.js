import { gql } from 'graphql-request';
import gqlClient from './graphQLClient';

/**
 *
 * @param {Object} query
 * @param {Integer} query.page
 * @param {Integer} query.limit
 *
 * @return {Promise<void>}
 */
import { ReferenceDataFragment } from './ReferenceDataAPI';

export async function advancedSearchApi(query = {}) {
  let response = await gqlClient.request(ProviderSearchQuery, query);

  return response.providersSearch;
}

const ReferenceDataLookupFragment = gql`
  fragment ReferenceDataLookup on ReferenceData {
    id
    name
    code
  }
`;

const ProviderLookupFragment = gql`
  fragment ProviderLookup on Provider {
    id
    status
    mode
    email
    mobile
    slug
    tagLine: tag_line
    legalFirstName: legal_first_name
    legalLastName: legal_last_name
    legalName: legal_name
    preferredContacts: preferred_contacts
    preferredFirstName: preferred_first_name
    name: program_name
    center: program_org_name
    description
    remote {
      available
      voice
      chat
      video
    }
    admission {
      type
      weekdays
      dates {
        start
        end
      }
    }
    distance(targetUnit: MI)
    providerTypes: provider_types {
      favorite
      providerType: provider_type {
        ...ReferenceData
      }
    }
    inPerson: in_person {
      available
    }
    availability {
      updated_at
      status
      afterHoursCrisisServices: after_hours_crisis_services
      intakeWaitDays: intake_wait_days
      hours {
        day
        start
        end
      }
    }
    photoUrl: photo_url
    paymentMethods: payment_methods {
      ...ReferenceDataLookup
    }
    paymentOptions: payment_options {
      ...ReferenceDataLookup
    }
    languageServices: language_services {
      ...ReferenceData
    }
    locations: locations {
      address {
        coordinates {
          coordinates: coordinates
        }
      }
      amenities {
        code
        category
      }
    }
    ageGroups: age_groups {
      ...ReferenceData
      ageStart: age_start
      ageEnd: age_end
    }
    treatsMedicallyUnstable: treats_medically_unstable
    canAssistWithDailyLiving: can_assist_with_daily_living
    treatsSuicidalIdeation: treats_suicidal_ideation
    notesCount: notes_count(organizationId: $organizationId)
    updatedAt: updated_at
    organizations {
      id
      subdomain
    }
  }
`;

const ProviderPaginationFragment = gql`
  fragment ProvidersPagination on ProviderPagination {
    items {
      ...ProviderLookup
    }
    pageInfo: page_info {
      perPage: per_page
      totalPages: total_pages
      totalPages: total_pages
      hasPrevPage: has_prev_page
      hasNextPage: has_next_page
      totalItems: total_items
    }
  }
`;

const ProviderSearchQuery = gql`
  query advancedSearch(
    $filter: ProvidersSearchInput
    $sort: ProvidersSearchOrder
    $page: UInt
    $limit: UInt
    $organizationId: ID
  ) {
    providersSearch(filter: $filter, sort: $sort, page: $page, limit: $limit) {
      ...ProvidersPagination
    }
  }
  ${ReferenceDataFragment}
  ${ReferenceDataLookupFragment}
  ${ProviderLookupFragment}
  ${ProviderPaginationFragment}
`;
