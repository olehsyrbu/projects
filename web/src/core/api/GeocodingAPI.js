import config from '../config';
import { getDistance } from 'geolib';

/**
 * Send address string to third-party (Google) api to
 * convert in longitude/latitude
 * @param {Object} context
 * @param {String} context.address - address string
 *
 * @return {Promise<Object[]>}
 */

const MAX_MINIMUM_DISTANCE = 20 * 1000;

export async function lookupLatLng({ address }) {
  let response = await fetch(
    `https://maps.googleapis.com/maps/api/geocode/json?address=${address}&components=country:US&key=${config.googleMapsApiKey}`,
  );
  response = await response.json();

  return response.results.map((result) => {
    let geometry = result.geometry;
    let addresses = result.address_components || [];

    const region = addresses.find((address) => {
      return address.types.includes('administrative_area_level_1');
    });

    const city = addresses.find((address) => {
      return address.types.includes('locality');
    });

    if (!geometry || !geometry.location) {
      return undefined;
    }

    let viewport = geometry.viewport;
    const distance = getDistance(
      {
        latitude: viewport.northeast.lat,
        longitude: viewport.northeast.lng,
      },
      {
        latitude: viewport.southwest.lat,
        longitude: viewport.southwest.lng,
      },
    );

    return {
      ...geometry,
      distance: {
        max: distance < MAX_MINIMUM_DISTANCE ? MAX_MINIMUM_DISTANCE : distance,
        min: 0,
      },
      region: region?.short_name,
      city: city?.short_name,
    };
  });
}
