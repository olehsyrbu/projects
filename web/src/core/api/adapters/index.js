export * from './programAdapter';
export * from './programDraftAdapter';
export * from './typesByLocationAdapter';
export * from './typesOnRemoteAdapter';
