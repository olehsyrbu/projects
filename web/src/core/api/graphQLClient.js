import { GraphQLClient } from 'graphql-request';
import config from '../config';

class Client extends GraphQLClient {
  constructor(url, options, config = {}) {
    super(url, options);
    this._config = config;
    this._errInterceptors = [];
  }

  setAuth(token) {
    this.setHeader('Authorization', `Bearer ${token}`);
  }

  removeAuth() {
    this.setHeader('Authorization', '');
  }

  async request(document, variables, requestHeaders) {
    let result = null;

    try {
      result = await super.request(document, variables, requestHeaders);
      return result;
    } catch (error) {
      this._runErrorInterceptors(error);
      throw error;
    }
  }

  _runErrorInterceptors(error) {
    // workaround for swr to not loose error because callbacks execution
    setTimeout(() => {
      this._errInterceptors.forEach((callback) => {
        callback(error);
      });
    });
  }

  addErrorInterceptor(fn) {
    this._errInterceptors.push(fn);
  }

  removeErrorInterceptor(fn) {
    const index = this._errInterceptors.findIndex(fn);
    if (index !== -1) {
      this._errInterceptors.splice(index, 1);
    }
  }
}

let client = new Client(config.api.baseUrl);

export default client;
