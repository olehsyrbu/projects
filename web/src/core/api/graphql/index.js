export * from './PageInfo';
export * from './TeamNotes';
export * from './UserProfile';
export * from './Programs';
