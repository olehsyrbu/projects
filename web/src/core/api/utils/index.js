export * from './parseAbilities';
export * from './filterInactiveEntities';
