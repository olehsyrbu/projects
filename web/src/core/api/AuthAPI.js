import { gql } from 'graphql-request';
import gqlClient from './graphQLClient';
import { UserProfileFragment } from './graphql';
import { parseAbilities } from '@/core/api/utils';

export async function register(params) {
  let response = await gqlClient.request(
    RegisterMutation,
    {
      user: {
        ...params,
      },
    },
    { Authorization: '' },
  );

  let {
    user,
    token: { auth },
  } = response.register;
  if (user && user.abilities) {
    user.abilities = parseAbilities(user.abilities);
  }

  return { user, token: auth };
}

export async function logIn(email, password) {
  let response = await gqlClient.request(LogInMutation, { email, password }, { Authorization: '' });
  let user = response.login.user;

  if (user && user.abilities) {
    user.abilities = parseAbilities(user.abilities);
  }

  return { user, token: response.login.token };
}

export async function logOut(token) {
  await gqlClient.request(LogOutMutation, null, { Authorization: `Bearer ${token}` });
}

export async function refreshToken() {
  // TODO: MIR-401 - enable refresh
  // let response = await gqlClient.request(RefreshTokenMutator);
  // gqlClient.setAuthToken(response.refreshToken.auth);
}

let RegisterMutation = gql`
  mutation registerUser($user: RegisterInput!) {
    register(user: $user) {
      user {
        ...UserProfile
      }
      token {
        auth
      }
    }
  }
  ${UserProfileFragment}
`;

let LogInMutation = gql`
  mutation login($email: Email!, $password: String!) {
    login(email: $email, password: $password) {
      user {
        ...UserProfile
      }
      token {
        auth
      }
    }
  }
  ${UserProfileFragment}
`;

let LogOutMutation = gql`
  mutation {
    logout {
      expires_at
    }
  }
`;
