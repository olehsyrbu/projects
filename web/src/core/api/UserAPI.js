import gqlClient from './graphQLClient';
import mixpanel from '@/core/mixpanel';
import { parseAbilities } from '@/core/api/utils';
import { UserProfileQuery, UserProfileMutation, DeleteUser } from './graphql';

export async function fetchUserProfile(token, client = gqlClient) {
  let header;

  if (token) {
    header = { Authorization: `Bearer ${token}` };
  }

  let response = await client.request(UserProfileQuery, undefined, header);
  let me = response.me;

  if (me && me.abilities) {
    me.abilities = parseAbilities(me.abilities);
  }

  return me;
}

export async function updateUserProfile(id, user) {
  let response = await gqlClient.request(UserProfileMutation, {
    id,
    user: {
      name: user.fullName,
      first_name: user.firstName,
      last_name: user.lastName,
      email: user.email,
      role: user.role,
      notification_email: user.notificationEmail,
    },
  });
  return response.updateUser;
}

export async function deleteUser(id) {
  const result = { err: null };

  try {
    await gqlClient.request(DeleteUser, {
      id,
    });
    mixpanel.track('Delete Self Account');
  } catch (err) {
    result.err = err;
    mixpanel.track('Delete Self Account Failure');
  }

  return result;
}

export async function switchRole(id, role) {
  return updateUserProfile(id, {
    role,
  });
}
