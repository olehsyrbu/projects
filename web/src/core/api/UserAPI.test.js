import { setupServer } from '@/tools/test-server/setupServer';
import { fetchUserProfile, switchRole } from '@/core/api/UserAPI';
import { userProfileFixtures } from '@/tools/test-fixtures';
import { parseAbilities } from '@/core/api/utils';
import { ROLE_NAMES } from '@/core/definitions';
import { userFixtures } from '@/tools/test-fixtures/userFixtures';

const server = setupServer();
const token = 'secret';

beforeAll(() => server.listen());
afterAll(() => server.close());
afterEach(() => server.resetHandlers());

function setupClient(returnValue) {
  return {
    request: vi.fn().mockReturnValue(Promise.resolve(returnValue)),
  };
}

it('call for user with no me', async () => {
  const client = setupClient({
    me: null,
  });

  const actual = await fetchUserProfile(token, client);

  expect(actual).toBeNull();
});

it('call for user query and return me', async () => {
  const client = setupClient({
    me: {
      abilities: [],
    },
  });

  const actual = await fetchUserProfile(token, client);

  expect(actual.abilities).toEqual([]);
});

it('replace abilities with parsed abilities', async () => {
  const client = setupClient({
    me: {
      abilities: [
        {
          subject: 'Provider',
          action: ['read', 'unaffiliate', 'delete', 'paginate'],
          conditions: ['{"owner.organization.id":"571c9349-502a-47ce-9aee-24f780891fa8"}'],
        },
        {
          subject: 'Invitation',
          action: ['read', 'write', 'delete', 'create', 'paginate'],
          conditions: ['{"owner.organization.id":"571c9349-502a-47ce-9aee-24f780891fa8"}'],
        },
      ],
    },
  });

  const actual = await fetchUserProfile(token, client);

  expect(actual.abilities).toMatchSnapshot();
});

it('call with default client', async () => {
  const actual = await fetchUserProfile();

  let expected = userProfileFixtures.defaultReferralCoordinatorProfileWithRawConditions();
  expect(actual).toEqual({
    ...expected,
    abilities: parseAbilities(expected.abilities),
  });
});

it('switches role for user', async () => {
  const actual = await switchRole('some-id', ROLE_NAMES.TEAM_MEMBER);
  expect(actual).toEqual(userFixtures.successResponseOnUserUpdate());
});
