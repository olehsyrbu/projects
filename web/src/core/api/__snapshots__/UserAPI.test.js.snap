// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replace abilities with parsed abilities 1`] = `
[
  {
    "action": [
      "read",
      "unaffiliate",
      "delete",
      "paginate",
    ],
    "conditions": {
      "owner.organization.id": "571c9349-502a-47ce-9aee-24f780891fa8",
    },
    "subject": "Provider",
  },
  {
    "action": [
      "read",
      "write",
      "delete",
      "create",
      "paginate",
    ],
    "conditions": {
      "owner.organization.id": "571c9349-502a-47ce-9aee-24f780891fa8",
    },
    "subject": "Invitation",
  },
]
`;
